### ISSUE TITLE
"Internal Error" : com.google.cloud.spark.bigquery.repackaged.io.grpc.StatusRuntimeException: INTERNAL: request failed: internal error

### ISSUE THREAD

**Issue by psrinuhp** [GoogleCloudDataproc/spark-bigquery-connector#433]
Hello All,

I am trying to read table metadata from BigQuery using below code.

Spark BigQuery Lib
```
spark-bigquery-with-dependencies_2.11-0.15.1-beta.jar
```

```
spark
.read
.bigquery("big-query-200217.132235582.__TABLES__")
.show(false)

```
I am getting below exception & Any kind of help is highly appreciated.

```
com.google.cloud.spark.bigquery.repackaged.com.google.api.gax.rpc.InternalException: com.google.cloud.spark.bigquery.repackaged.io.grpc.StatusRuntimeException: INTERNAL: request failed: internal error
	at com.google.cloud.spark.bigquery.repackaged.com.google.api.gax.rpc.ApiExceptionFactory.createException(ApiExceptionFactory.java:67)
	at com.google.cloud.spark.bigquery.repackaged.com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:72)
	at com.google.cloud.spark.bigquery.repackaged.com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:60)
	at com.google.cloud.spark.bigquery.repackaged.com.google.api.gax.grpc.GrpcExceptionCallable$ExceptionTransformingFuture.onFailure(GrpcExceptionCallable.java:97)
	at com.google.cloud.spark.bigquery.repackaged.com.google.api.core.ApiFutures$1.onFailure(ApiFutures.java:68)
	at com.google.cloud.spark.bigquery.repackaged.com.google.common.util.concurrent.Futures$CallbackListener.run(Futures.java:1074)
	at com.google.cloud.spark.bigquery.repackaged.com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:30)
	at com.google.cloud.spark.bigquery.repackaged.com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:1213)
	at com.google.cloud.spark.bigquery.repackaged.com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:983)
	at com.google.cloud.spark.bigquery.repackaged.com.google.common.util.concurrent.AbstractFuture.setException(AbstractFuture.java:771)
	at com.google.cloud.spark.bigquery.repackaged.io.grpc.stub.ClientCalls$GrpcFuture.setException(ClientCalls.java:563)
	at com.google.cloud.spark.bigquery.repackaged.io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose(ClientCalls.java:533)
	at com.google.cloud.spark.bigquery.repackaged.io.grpc.internal.DelayedClientCall$DelayedListener$3.run(DelayedClientCall.java:463)
	at com.google.cloud.spark.bigquery.repackaged.io.grpc.internal.DelayedClientCall$DelayedListener.delayOrExecute(DelayedClientCall.java:427)
	at com.google.cloud.spark.bigquery.repackaged.io.grpc.internal.DelayedClientCall$DelayedListener.onClose(DelayedClientCall.java:460)
	at com.google.cloud.spark.bigquery.repackaged.io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:553)
	at com.google.cloud.spark.bigquery.repackaged.io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:68)
	at com.google.cloud.spark.bigquery.repackaged.io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:739)
	at com.google.cloud.spark.bigquery.repackaged.io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:718)
	at com.google.cloud.spark.bigquery.repackaged.io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at com.google.cloud.spark.bigquery.repackaged.io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
	Suppressed: com.google.cloud.spark.bigquery.repackaged.com.google.api.gax.rpc.AsyncTaskException: Asynchronous task failed
		at com.google.cloud.spark.bigquery.repackaged.com.google.api.gax.rpc.ApiExceptions.callAndTranslateApiException(ApiExceptions.java:57)
		at com.google.cloud.spark.bigquery.repackaged.com.google.api.gax.rpc.UnaryCallable.call(UnaryCallable.java:112)
		at com.google.cloud.spark.bigquery.repackaged.com.google.cloud.bigquery.storage.v1.BigQueryReadClient.createReadSession(BigQueryReadClient.java:230)
		at com.google.cloud.spark.bigquery.direct.DirectBigQueryRelation.buildScan(DirectBigQueryRelation.scala:134)
		at org.apache.spark.sql.execution.datasources.DataSourceStrategy$$anonfun$11.apply(DataSourceStrategy.scala:277)
		at org.apache.spark.sql.execution.datasources.DataSourceStrategy$$anonfun$11.apply(DataSourceStrategy.scala:277)
		at org.apache.spark.sql.execution.datasources.DataSourceStrategy$$anonfun$pruneFilterProject$1.apply(DataSourceStrategy.scala:321)
		at org.apache.spark.sql.execution.datasources.DataSourceStrategy$$anonfun$pruneFilterProject$1.apply(DataSourceStrategy.scala:320)
		at org.apache.spark.sql.execution.datasources.DataSourceStrategy.pruneFilterProjectRaw(DataSourceStrategy.scala:401)
		at org.apache.spark.sql.execution.datasources.DataSourceStrategy.pruneFilterProject(DataSourceStrategy.scala:316)
		at org.apache.spark.sql.execution.datasources.DataSourceStrategy.apply(DataSourceStrategy.scala:273)
		at org.apache.spark.sql.catalyst.planning.QueryPlanner$$anonfun$1.apply(QueryPlanner.scala:62)
		at org.apache.spark.sql.catalyst.planning.QueryPlanner$$anonfun$1.apply(QueryPlanner.scala:62)
		at scala.collection.Iterator$$anon$12.nextCur(Iterator.scala:434)
		at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:440)
		at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:439)
		at org.apache.spark.sql.catalyst.planning.QueryPlanner.plan(QueryPlanner.scala:92)
		at org.apache.spark.sql.catalyst.planning.QueryPlanner$$anonfun$2$$anonfun$apply$2.apply(QueryPlanner.scala:77)
		at org.apache.spark.sql.catalyst.planning.QueryPlanner$$anonfun$2$$anonfun$apply$2.apply(QueryPlanner.scala:74)
		at scala.collection.TraversableOnce$$anonfun$foldLeft$1.apply(TraversableOnce.scala:157)
		at scala.collection.TraversableOnce$$anonfun$foldLeft$1.apply(TraversableOnce.scala:157)
		at scala.collection.Iterator$class.foreach(Iterator.scala:893)
		at scala.collection.AbstractIterator.foreach(Iterator.scala:1336)
		at scala.collection.TraversableOnce$class.foldLeft(TraversableOnce.scala:157)
		at scala.collection.AbstractIterator.foldLeft(Iterator.scala:1336)
		at org.apache.spark.sql.catalyst.planning.QueryPlanner$$anonfun$2.apply(QueryPlanner.scala:74)
		at org.apache.spark.sql.catalyst.planning.QueryPlanner$$anonfun$2.apply(QueryPlanner.scala:66)
		at scala.collection.Iterator$$anon$12.nextCur(Iterator.scala:434)
		at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:440)
		at org.apache.spark.sql.catalyst.planning.QueryPlanner.plan(QueryPlanner.scala:92)
		at org.apache.spark.sql.execution.QueryExecution.sparkPlan$lzycompute(QueryExecution.scala:84)
		at org.apache.spark.sql.execution.QueryExecution.sparkPlan(QueryExecution.scala:80)
		at org.apache.spark.sql.execution.QueryExecution.executedPlan$lzycompute(QueryExecution.scala:89)
		at org.apache.spark.sql.execution.QueryExecution.executedPlan(QueryExecution.scala:89)
		at org.apache.spark.sql.Dataset.withAction(Dataset.scala:2832)
		at org.apache.spark.sql.Dataset.head(Dataset.scala:2153)
		at org.apache.spark.sql.Dataset.take(Dataset.scala:2366)
		at org.apache.spark.sql.Dataset.showString(Dataset.scala:245)
		at org.apache.spark.sql.Dataset.show(Dataset.scala:646)
		at org.apache.spark.sql.Dataset.show(Dataset.scala:623)
		
Caused by: com.google.cloud.spark.bigquery.repackaged.io.grpc.StatusRuntimeException: INTERNAL: request failed: internal error
	at com.google.cloud.spark.bigquery.repackaged.io.grpc.Status.asRuntimeException(Status.java:535)
	... 17 more
```

**Comment by kmjung** [comment]
The __TABLES__ meta-table is not supported by the storage API (like the __DATASET__ and __TABLES_SUMMARY__ meta-tables), and at the moment there are no plans to change this. What is your use case here?

**Comment by kmjung** [comment]
I have verified on our side that this is a regression and have filed a bug to address the underlying issue. However, the `__TABLES__` meta-table is not the recommended approach for fetching table information in the first place; does the `TABLES` table in `INFORMATION_SCHEMA` have the information you need? https://cloud.google.com/bigquery/docs/information-schema-tables

**Comment by chajath** [comment]
@kmjung I'm in the same boat. Having trouble querying `project_id.data_set.INFORMATION_SCHEMA.TABLES` since it breaks some hard-wired assumption about the table naming. I get the following error:

```
[error] {
[error]   "code" : 400,
[error]   "errors" : [ {
[error]     "domain" : "global",
[error]     "message" : "Invalid project ID 'project_id:data_set'. Project IDs must contain 6-63 lowercase letters, digits, or dashes. Some project IDs also include domain name separated by a colon. IDs must start with a letter and may not end with a dash.",
[error]     "reason" : "invalid"
[error]   } ],
[error]   "message" : "Invalid project ID 'project_id:data_set'. Project IDs must contain 6-63 lowercase letters, digits, or dashes. Some project IDs also include domain name separated by a colon. IDs must start with a letter and may not end with a dash.",
[error]   "status" : "INVALID_ARGUMENT"
[error] }
```

Is there a way to work around this to query `INFORMATION_SCHEMA.TABLES` ?

**Comment by kmjung** [comment]
Sorry, to be clear: the `__TABLES__` meta-table and the `TABLES` view in `INFORMATION_SCHEMA` are separate entities in BigQuery and the underlying issues are different. @chajath can you file a new issue here?

**Comment by psrinuhp** [comment]
Thank You @kmjung.

Use case is to get the latest `table id`, `total rows` & `size` for given dataset & I am fetching these details from `__TABLES__`

also I have tried following code but I am getting different error which @chajath has mentioned in above.
```
spark.read
.option("parentProjectId","")
.option("projectId","")
.option("dataset","")
.bigquery("INFORMATION_SCHEMA.TABLES")
```
```
spark.read.bigquery("<projectId>:<dataset>.INFORMATION_SCHEMA.TABLES")
```

**Comment by kmjung** [comment]
Note that for row counts and size, you  may need to query the INFORMATION_SCHEMA.PARTITIONS view as well.

**Comment by psrinuhp** [comment]
I have executed below code & getting error `java.lang.RuntimeException: Option table required.`

Do i need to add any config properties to fix below issue ?

```
scala> spark.conf.set("viewsEnabled","true")
scala> spark.conf.set("materializationDataset","<dataset>")
```
```
scala> val df = spark.read.format("bigquery").load("select * from `<projectId>.<dataset>.INFORMATION_SCHEMA.TABLES`")
java.lang.RuntimeException: Option table required.
  at scala.sys.package$.error(package.scala:27)
  at com.google.cloud.spark.bigquery.SparkBigQueryOptions$$anonfun$getRequiredOption$1.apply(SparkBigQueryOptions.scala:198)
  at com.google.cloud.spark.bigquery.SparkBigQueryOptions$$anonfun$getRequiredOption$1.apply(SparkBigQueryOptions.scala:198)
  at scala.Option.getOrElse(Option.scala:121)
  at com.google.cloud.spark.bigquery.SparkBigQueryOptions$.getRequiredOption(SparkBigQueryOptions.scala:198)
  at com.google.cloud.spark.bigquery.SparkBigQueryOptions$.apply(SparkBigQueryOptions.scala:112)
  at com.google.cloud.spark.bigquery.BigQueryRelationProvider.createSparkBigQueryOptions(BigQueryRelationProvider.scala:114)
  at com.google.cloud.spark.bigquery.BigQueryRelationProvider.createRelationInternal(BigQueryRelationProvider.scala:52)
  at com.google.cloud.spark.bigquery.BigQueryRelationProvider.createRelation(BigQueryRelationProvider.scala:37)
  at org.apache.spark.sql.execution.datasources.DataSource.resolveRelation(DataSource.scala:306)
  at org.apache.spark.sql.DataFrameReader.load(DataFrameReader.scala:178)
  at org.apache.spark.sql.DataFrameReader.load(DataFrameReader.scala:156)
  ... 50 elided

```

**Comment by kmjung** [comment]
Hmm, I'm not sure at a glance whether the query syntax is supported in v2.11 of the connector -- @davidrabinowitz do you know?

**Comment by kmjung** [comment]
Correction: this is not supported in v0.15.1-beta. Per the release notes, support for direct queries was added in 0.19.0: https://github.com/GoogleCloudDataproc/spark-bigquery-connector/issues/247

**Comment by psrinuhp** [comment]
> I have verified on our side that this is a regression and have filed a bug to address the underlying issue. However, the `__TABLES__` meta-table is not the recommended approach for fetching table information in the first place; does the `TABLES` table in `INFORMATION_SCHEMA` have the information you need? https://cloud.google.com/bigquery/docs/information-schema-tables

Thank You @kmjung. 

We are using `spark 2.2.0` & `spark-bigquery-with-dependencies_2.11-0.15.1-beta.jar` version in our cluster.

Any idea when this bug will be fixed  ?

**Comment by kmjung** [comment]
It may take up to 2 weeks before we can roll out a fix here. You can follow https://issuetracker.google.com/issues/194114043 for the status of the fix.

With that said -- if it's possible to do so, I strongly recommend that you consider moving to the new PARTITIONS view in the INFORMATION_SCHEMA table. The `__TABLES__` meta-table is undocumented and officially unsupported, and the PARTITIONS view won't have the scalability and reliability issues that `__TABLES__` does.

**Comment by davidrabinowitz** [comment]
@psrinuhp Please upgrade to the latest connector version, 0.21.1. After you do that, you would be able to run the  following:
```
spark.conf.set("viewsEnabled","true")
spark.conf.set("materializationDataset","<dataset>")
val tablesDF = spark.read.format("bigquery").load("select * from `<projectId>.<dataset>.__TABLES__`")
```

**Comment by psrinuhp** [comment]
Thank You @davidrabinowitz . we are currently using spark `2.2.0` version. will this latest bigquery connector work in spark 2.2.0 ?

**Comment by davidrabinowitz** [comment]
It has been tested on Spark 2.3, and we are not using any new 2.3+ features to the best of my knowledge. In that respect it is not very different from version 0.15 you are currently using.

**Comment by kmjung** [comment]
@psrinuhp if you're able to resolve this issue using a more recent release of the connector, please update this issue with the solution.

**Comment by psrinuhp** [comment]
@kmjung , I am getting following exception with more recent release of the bigquery connector.

```
Spark Version - 2.2.0
BigQuery Connector Version - 0.21.1
```

```
spark-shell --jars spark-bigquery-with-dependencies_2.11-0.21.1.jar
```
```
scala> spark.conf.set("viewsEnabled","true")
scala> spark.conf.set("materializationDataset","<dataset>")
scala> val df = spark.read.format("bigquery").load("select * from `<projectId>.<dataset>.INFORMATION_SCHEMA.TABLES`")
df: org.apache.spark.sql.DataFrame = [table_catalog: string, table_schema: string ... 5 more fields]
```

scala> df.show(false)
```
21/07/21 18:11:22 ERROR Executor: Exception in task 0.0 in stage 0.0 (TID 0)
java.lang.BootstrapMethodError: java.lang.NoClassDefFoundError: org/apache/spark/sql/vectorized/ColumnVector
	at com.google.cloud.spark.bigquery.ArrowBinaryIterator.toArrowRows(ArrowBinaryIterator.java:103)
	at com.google.cloud.spark.bigquery.ArrowBinaryIterator.hasNext(ArrowBinaryIterator.java:89)
	at scala.collection.convert.Wrappers$JIteratorWrapper.hasNext(Wrappers.scala:42)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:438)
	at org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIterator.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anonfun$8$$anon$1.hasNext(WholeStageCodegenExec.scala:395)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$2.apply(SparkPlan.scala:234)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$2.apply(SparkPlan.scala:228)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$25.apply(RDD.scala:827)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$25.apply(RDD.scala:827)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:38)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:323)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:287)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:87)
	at org.apache.spark.scheduler.Task.run(Task.scala:108)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:335)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NoClassDefFoundError: org/apache/spark/sql/vectorized/ColumnVector
	at java.lang.ClassLoader.defineClass1(Native Method)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:756)
	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
	at java.net.URLClassLoader.defineClass(URLClassLoader.java:468)
	at java.net.URLClassLoader.access$100(URLClassLoader.java:74)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:369)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:363)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.URLClassLoader.findClass(URLClassLoader.java:362)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
	... 21 more
Caused by: java.lang.ClassNotFoundException: org.apache.spark.sql.vectorized.ColumnVector
	at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
	... 32 more
```

**Comment by davidrabinowitz** [comment]
The connector has switched to use Apache Arrow as the default wire format as it is more performant.  As the support for this class had been added in spark 2.3, you can try to add the following configuration to your spark application:
```
spark.conf.set("readDataFormat", "AVRO")
```
